components/registration-framework/org.wso2.carbon.identity.user.registration.engine/src/main/java/org/wso2/carbon/identity/user/registration/engine/UserRegistrationFlowService.java

```diff
@@ -0,0 +1,86 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.user.registration.engine;
+
+import static org.wso2.carbon.identity.user.registration.engine.util.Constants.SELECTED_ACTION;
+import static org.wso2.carbon.identity.user.registration.engine.util.Constants.STATUS_COMPLETE;
+import java.util.Map;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.user.registration.engine.model.RegistrationContext;
+import org.wso2.carbon.identity.user.registration.engine.model.RegistrationStep;
+import org.wso2.carbon.identity.user.registration.engine.util.Constants;
+import org.wso2.carbon.identity.user.registration.engine.util.RegistrationFlowEngineUtils;
+import org.wso2.carbon.identity.user.registration.mgt.exception.RegistrationFrameworkException;
+
+/**
+ * Service class to handle the user registration flow.
+ */
+public class UserRegistrationFlowService {
+
+    private static final Log LOG = LogFactory.getLog(UserRegistrationFlowService.class);
+    private static final UserRegistrationFlowService instance = new UserRegistrationFlowService();
+
+    private UserRegistrationFlowService() {
+
+    }
+
+    public static UserRegistrationFlowService getInstance() {
+
+        return instance;
+    }
+
+    /**
+     * Initiates the registration flow for the given tenant domain.
+     *
+     * @param tenantDomain Tenant domain.
+     * @return RegistrationStep.
+     * @throws RegistrationFrameworkException if something goes wrong while initiating the registration flow.
+     */
+    public RegistrationStep initiateDefaultRegistrationFlow(String tenantDomain) throws RegistrationFrameworkException {
+
+        RegistrationContext context = RegistrationFlowEngineUtils.initiateContext(tenantDomain);
+        RegistrationStep step = RegistrationFlowEngine.getInstance().execute(context);
+        RegistrationFlowEngineUtils.addRegContextToCache(context);
+        return step;
+    }
+
+    /**
+     * Continues the registration flow for the given flow ID with the provided user inputs.
+     *
+     * @param flowId Flow ID.
+     * @param inputs User inputs.
+     * @return ExecutionState.
+     * @throws RegistrationFrameworkException if something goes wrong while continuing the registration flow.
+     */
+    public RegistrationStep continueFlow(String flowId, String actionId, Map<String, String> inputs)
+            throws RegistrationFrameworkException {
+
+        RegistrationContext context = RegistrationFlowEngineUtils.retrieveRegContextFromCache(flowId);
+        context.getUserInputData().putAll(inputs);
+        context.getUserInputData().put(SELECTED_ACTION, actionId);
```

Comment:
Can we treat this as a user input?

Reply:
Introduced a new attribute to the context to store the action id in each iteration.